### PR TITLE
gradle: add package tests

### PR DESCRIPTION
**Pull request by liff**
## Description of changes

Add simple package tests for Gradle.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by liff** [comment]
@lorenzleutgeb sorry to bother, but pinging in case you have time to take a peek.

**Comment by lorenzleutgeb** [comment]
@liff I am on vacation until incl. 2024-02-12. I like the change but don't count on me testing it before then, sorry.

**Comment by liff** [comment]
Oh even more sorry I bothered you then, have a great rest of your vacation!

File: pkgs/development/tools/build-managers/gradle/default.nix [hunk 1]
```diff
@@ -99,6 +108,24 @@ rec {
...
 
+      passthru.tests = {
+        version = testers.testVersion { package = finalAttrs.finalPackage; };
+
+        java-application = testers.testEqualContents {
+          assertion = "can build and run a trivial Java application";
+          expected = writeText "expected" ''
+            hello
+          '';
+          actual = runCommand "java-application" {
```
Comment (lorenzleutgeb) [inline pkgs/development/tools/build-managers/gradle/default.nix:119]:
```suggestion
          actual = runCommand "actual" {
```

File: pkgs/development/tools/build-managers/gradle/default.nix [hunk 2]
```diff
@@ -99,6 +108,24 @@ rec {
...
       '';
 
+      passthru.tests = {
+        version = testers.testVersion { package = finalAttrs.finalPackage; };
+
+        java-application = testers.testEqualContents {
+          assertion = "can build and run a trivial Java application";
+          expected = writeText "expected" ''
+            hello
+          '';
```
Comment (lorenzleutgeb) [inline pkgs/development/tools/build-managers/gradle/default.nix:118]:
```suggestion
          expected = writeText "expected" "hello\n";
```

**Review by lorenzleutgeb — APPROVED** [review]
Awesome, thanks! Builds on my machine (incl. tests). Two minor suggestions.

**Comment by liff** [comment]
Is this good to approve?

**Review by lorenzleutgeb — APPROVED** [review]

File: pkgs/development/tools/build-managers/gradle/default.nix [hunk 3]
```diff
@@ -99,6 +108,22 @@ rec {
         echo ${ncurses6} >> $out/nix-support/manual-runtime-dependencies
       '';
 
+      passthru.tests = {
```
Comment (SuperSandro2000) [inline pkgs/development/tools/build-managers/gradle/default.nix:111]: fails on darwin

```
FAILURE: Build failed with an exception.
* What went wrong:
Gradle could not start your build.
> Could not initialize native services.
   > Failed to load native library 'libnative-platform.dylib' for Mac OS X x86_64.
* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.
gradle --version returned a non-zero exit code.
FAILURE: Build failed with an exception.
* What went wrong:
Gradle could not start your build.
> Could not initialize native services.
   > Failed to load native library 'libnative-platform.dylib' for Mac OS X x86_64.
* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.
error: builder for '/nix/store/pdfs37wfivmii12mqmx95xp6vlzn53qz-actual.drv' failed with exit code 1;
```

Comment (liff) [reply pkgs/development/tools/build-managers/gradle/default.nix:111]: That seems to suggest that Gradle doesn’t work on macOS at all. Guess I’ll have to try to find a macOS to test with.

Comment (liff) [reply pkgs/development/tools/build-managers/gradle/default.nix:111]: Turns out Gradle wants to extract the library to `$HOME/native`, which is not writable on macOS, apparently. Will try to figure out how to work around that.

Comment (liff) [reply pkgs/development/tools/build-managers/gradle/default.nix:111]: Should work on macOS now.

**Pull request merged**